File: store-core/console-ui/src/app/api/catalog/product-image.service.ts
import {Injectable, inject} from '@angular/core';
import {Observable} from 'rxjs';

import {CrudService} from '@core/http/crud.service';
import type {PersistableProductImage, ReadableImage} from '@models/catalog';

const CATALOG_API_BASE = '/spg/catalog/api/v1';

/**
 * Verified against `catalog-service/api/v1/ProductImageApi.java`.
 *
 * A product's gallery — references into the content service's media library.
 *
 * **Catalog does not accept uploads.** The bytes go to the media library, where they are
 * deduplicated, measured, given alt text and protected from deletion while something still shows
 * them; catalog stores the asset ids. That is what closed two holes this comment used to describe:
 * a product's default image could be seen but not changed, and an image could be deleted from under
 * a live listing with nothing noticing.
 */
@Injectable({providedIn: 'root'})
export class ProductImageService {
  private readonly crudService = inject(CrudService);

  /**
   * Every image on a product, in the order the pod holds them.
   *
   * A **public** read — `/product/{id}/images`, no `private` — because the storefront gallery uses
   * the same endpoint. Verified against the mapping; not a missing segment.
   */
  images(productId: number): Observable<ReadableImage[]> {
    return this.crudService.get(`${CATALOG_API_BASE}/product/${productId}/images`);
  }

  /**
   * Appends images after the ones the product already has.
   *
   * An asset id that is not this store's is refused with `CATALOG.PRODUCT_IMAGE.ASSET_UNKNOWN` and
   * nothing is written.
   */
  attach(productId: number, items: readonly PersistableProductImage[]): Observable<ReadableImage[]> {
    return this.crudService.post(`${CATALOG_API_BASE}/private/product/${productId}/images`, items);
  }

  /**
   * Replaces the whole gallery: order is the list order, and the item flagged default wins.
   *
   * Sent whole because a reorder that arrives one move at a time leaves gaps and ties that the
   * storefront resolves arbitrarily. This is also how the default image is changed.
   */
  replace(productId: number, items: readonly PersistableProductImage[]): Observable<ReadableImage[]> {
    return this.crudService.put(`${CATALOG_API_BASE}/private/product/${productId}/images`, items);
  }

  /** Detaches an image. The asset stays in the library, where other products may still use it. */
  remove(productId: number, imageId: number): Observable<void> {
    return this.crudService.delete(
      `${CATALOG_API_BASE}/private/product/${productId}/image/${imageId}`,
    );
  }
}

File: store-pod/catalog/catalog-core/src/main/java/com/asrevo/cvhome/catalog/services/image/ProductImageService.java
package com.asrevo.cvhome.catalog.services.image;

import java.util.List;

import com.asrevo.cvhome.catalog.entity.Product;
import com.asrevo.cvhome.catalog.errors.ProductImageAssetUnknownException;
import com.asrevo.cvhome.catalog.errors.ProductImageNotFoundException;
import com.asrevo.cvhome.catalog.errors.ProductNotFoundException;
import com.asrevo.cvhome.catalog.model.product.PersistableProductImage;
import com.asrevo.cvhome.catalog.model.product.ReadableImage;
import com.asrevo.cvhome.commons.domain.StoreMerchantId;
import com.asrevo.cvhome.errors.RemoteServiceTimeoutException;
import com.asrevo.cvhome.errors.RemoteServiceUnavailableException;

/**
 * A product's gallery: references into the content service's media library, in display order.
 *
 * <p>
 * Uploads are not catalog's job any more. The seller puts bytes in the media library and catalog stores the
 * asset ids, which is what lets an image carry alt text and dimensions, be reused across products, and be
 * protected from deletion while something still shows it.
 * </p>
 */
public interface ProductImageService {

    List<ReadableImage> list(StoreMerchantId store, Long productId) throws ProductNotFoundException;

    /**
     * Appends images after the ones the product already has.
     *
     * @throws ProductImageAssetUnknownException  an asset id is not in this store's media library
     * @throws RemoteServiceUnavailableException  content could not be reached — nothing is written, because a row
     *                                            with no url is worse than a failed save
     * @throws RemoteServiceTimeoutException      content did not answer in time
     */
    List<ReadableImage> attach(StoreMerchantId store, Long productId, List<PersistableProductImage> items)
            throws ProductNotFoundException, ProductImageAssetUnknownException, RemoteServiceUnavailableException,
            RemoteServiceTimeoutException;

    /**
     * Replaces the whole gallery: order is the list order, and the item flagged default wins. Sent whole because
     * a reorder that arrives one move at a time leaves gaps and ties the storefront resolves arbitrarily.
     */
    List<ReadableImage> replace(StoreMerchantId store, Long productId, List<PersistableProductImage> items)
            throws ProductNotFoundException, ProductImageAssetUnknownException, RemoteServiceUnavailableException,
            RemoteServiceTimeoutException;

    /**
     * Detaches an image from the product. The asset itself stays in the library.
     */
    void delete(StoreMerchantId store, Long productId, Long imageId) throws ProductImageNotFoundException;

    /**
     * Releases a product's hold on its assets before the product is deleted; the rows go with it.
     */
    void forget(Product product);
}
